RecoEgamma/EgammaTools/plugins/EGRegressionModifierV3.cc

```diff
@@ -0,0 +1,364 @@
+    EpCombinationTool epComb;
+  };
+
+  struct PhoRegs {
+    PhoRegs(const edm::ParameterSet& iConfig);
+    void setEventContent(const edm::EventSetup& iSetup);
+    EgammaRegressionContainer ecalOnlyMean;
+    EgammaRegressionContainer ecalOnlySigma;    
+  };
+
+  EGRegressionModifierV3(const edm::ParameterSet& conf, edm::ConsumesCollector& cc);
+  ~EGRegressionModifierV3() override;
+    
+  void setEvent(const edm::Event&) final;
+  void setEventContent(const edm::EventSetup&) final;
+  
+  void modifyObject(reco::GsfElectron&) const final;
+  void modifyObject(reco::Photon&) const final;
+  
+  // just calls reco versions
+  void modifyObject(pat::Electron&) const final; 
+  void modifyObject(pat::Photon&) const final;
+
+private:
+  std::array<float,32> getRegData(const reco::GsfElectron& ele)const;
+  std::array<float,32> getRegData(const reco::Photon& pho)const;
+  void getSeedCrysCoord(const reco::CaloCluster& clus,int& iEtaOrX,int& iPhiOrY)const;
+
+  std::unique_ptr<EleRegs> eleRegs_;
+  std::unique_ptr<PhoRegs> phoRegs_;
+  
+  float rhoValue_;
+  edm::EDGetTokenT<double> rhoToken_;
+
+  bool useClosestToCentreSeedCrysDef_;
+  edm::ESHandle<CaloGeometry> caloGeomHandle_;
+
+};
+
+DEFINE_EDM_PLUGIN(ModifyObjectValueFactory,
+		  EGRegressionModifierV3,
+		  "EGRegressionModifierV3");
+
+EGRegressionModifierV3::EGRegressionModifierV3(const edm::ParameterSet& conf, 
+					       edm::ConsumesCollector& cc) :
+  ModifyObjectValueBase(conf),
+  rhoValue_(0.),
+  rhoToken_(cc.consumes<double>(conf.getParameter<edm::InputTag>("rhoTag"))),
+  useClosestToCentreSeedCrysDef_(conf.getParameter<bool>("useClosestToCentreSeedCrysDef"))
+{   
+  if(conf.exists("eleRegs")) {
+    eleRegs_ = std::make_unique<EleRegs>(conf.getParameter<edm::ParameterSet>("eleRegs"));
+  }
+  if(conf.exists("phoRegs")) {
+    phoRegs_ = std::make_unique<PhoRegs>(conf.getParameter<edm::ParameterSet>("phoRegs"));
+  }
+}
+
+EGRegressionModifierV3::~EGRegressionModifierV3(){}
+
+void EGRegressionModifierV3::setEvent(const edm::Event& evt)
+{
+  edm::Handle<double> rhoHandle;
+  evt.getByToken(rhoToken_, rhoHandle);
+  rhoValue_ = *rhoHandle;
+}
+
+void EGRegressionModifierV3::setEventContent(const edm::EventSetup& iSetup)
+{
+  if(eleRegs_) eleRegs_->setEventContent(iSetup);
+  if(phoRegs_) phoRegs_->setEventContent(iSetup);
+  if(useClosestToCentreSeedCrysDef_) iSetup.get<CaloGeometryRecord>().get(caloGeomHandle_);
+}
+
+void EGRegressionModifierV3::modifyObject(reco::GsfElectron& ele) const
+{
+  //check if we have specified an electron regression correction and
+  //return the object unmodified if so
+  if(!eleRegs_) return;
+
+  const reco::SuperClusterRef& superClus = ele.superCluster();
+  
+  // skip HGCAL for now
+  if( superClus->seed()->seed().det() == DetId::Forward ) return;
+
+  const int numberOfClusters =  superClus->clusters().size();
+  const bool missingClusters = !superClus->clusters()[numberOfClusters-1].isAvailable();
```

Comment:
so size 0 should never happen but I think its possible to happen if somebody screws up creating a supercluster in some future bit of code. So i'll add a check.  

If one cluster is available, then they all should be available, hence we just need to check one. No good reason for checking the back rather than front, I choose back incase somebody did something cute with the seed clusters (aka the front). Will make the change as suggested.

Reply:
so I found out why I did it the rather odd way I did it :)

Turns out I had a reason. Its a PtrVector and has no method "back".  Still did the size 0 check just in case.  So left as is but with a check for  size=0.